File: src/dorkbox/util/cab/structure/CabFolderEntry.java
/*
 * Copyright 2012 dorkbox, llc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dorkbox.util.cab.structure;

import java.io.IOException;
import java.io.InputStream;

import dorkbox.util.bytes.LittleEndian;

public final class CabFolderEntry implements CabConstants {
    /** offset of the first CFDATA block in this folder, 4bytes */
    public long coffCabStart;
    /** number of CFDATA blocks in this folder, 2bytes */
    public int  cCFData;
    /** compression type indicator , 2bytes */
    public int  compressionMethod = 0;

    public int getCompressionMethod() {
        return this.compressionMethod & 0xF;
    }

    public void read(InputStream input) throws IOException {
        this.coffCabStart = LittleEndian.UInt_.from(input).longValue();
        this.cCFData = LittleEndian.UShort_.from(input).intValue();
        this.compressionMethod = LittleEndian.UShort_.from(input).intValue();
    }

    public int getCompressionWindowSize() {
        if (this.compressionMethod == COMPRESSION_TYPE_NONE) {
            return 0;
        }
        if (this.compressionMethod == COMPRESSION_TYPE_MSZIP) {
            return 16;
        }
        return (this.compressionMethod & 0x1F00) >>> 8;
    }

    public String compressionToString() {
        switch (getCompressionMethod()) {
            default :
                return "UNKNOWN";
            case 0 :
                return "NONE";
            case 1 :
                return "MSZIP";
            case 2 :
                return "QUANTUM:" + Integer.toString(getCompressionWindowSize());
            case 3 :
        }
        return "LZX:" + Integer.toString(getCompressionWindowSize());
    }

    public void setCompression(int a, int b) {
        this.compressionMethod = b << 8 | a;
    }
}

File: src/dorkbox/cabParser/structure/CabFolderEntry.kt
/*
 * Copyright 2023 dorkbox, llc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dorkbox.cabParser.structure

import dorkbox.bytes.LittleEndian
import java.io.IOException
import java.io.InputStream

class CabFolderEntry : CabConstants {
    /** offset of the first CFDATA block in this folder, 4bytes  */
    var coffCabStart: Long = 0

    /** number of CFDATA blocks in this folder, 2bytes  */
    var cCFData = 0

    /** compression type indicator , 2bytes  */
    var compressionMethod = 0

    @Throws(IOException::class)
    fun read(input: InputStream) {
        coffCabStart = LittleEndian.UInt_.from(input).toLong()
        cCFData = LittleEndian.UShort_.from(input).toInt()
        compressionMethod = LittleEndian.UShort_.from(input).toInt()
    }

    val compressionWindowSize: Int
        get() {
            if (compressionMethod == CabConstants.COMPRESSION_TYPE_NONE) {
                return 0
            }
            return if (compressionMethod == CabConstants.COMPRESSION_TYPE_MSZIP) {
                16
            }
            else compressionMethod and 0x1F00 ushr 8
        }

    fun compressionToString(): String {
        when (compressionMethod) {
            0    -> return "NONE"
            1    -> return "MSZIP"
            2    -> return "QUANTUM:" + compressionWindowSize.toString()
            3    -> {}
            else -> return "UNKNOWN"
        }
        return "LZX:" + compressionWindowSize.toString()
    }

    fun setCompression(a: Int, b: Int) {
        compressionMethod = b shl 8 or a
    }
}
